crates/communication/src/server/server.rs

```diff
@@ -0,0 +1,123 @@
+
+use framework::{Reader, Writer};
+use serialize_hierarchy::SerializeHierarchy;
+use tokio::{
+    runtime::{self, Runtime},
+    sync::{
+        mpsc::{channel, Sender},
+        oneshot, Notify,
+    },
+};
+use tokio_util::sync::CancellationToken;
+
+use crate::server::databases::router::router;
+
+use super::{
+    acceptor::{acceptor, AcceptError},
+    databases::{provider::provider, Request},
+};
+
+#[derive(Debug, thiserror::Error)]
+pub enum StartError {
+    #[error("error while accepting connections")]
+    AcceptError(AcceptError),
+    #[error("one or more tasks encountered an error")]
+    TasksErrored(Vec<StartError>),
+    #[error("thread not started")]
+    ThreadNotStarted(io::Error),
+    #[error("runtime not started")]
+    RuntimeNotStarted(io::Error),
+}
+
+pub struct Server {
+    runtime: Arc<Runtime>,
+    databases_sender: Sender<Request>,
+}
+
+impl Server {
+    pub fn start(keep_running: CancellationToken) -> Result<Self, StartError> {
+        let (runtime_sender, runtime_receiver) = oneshot::channel();
+
+        let join_handle = thread::Builder::new()
+            .name("communication".to_string())
+            .spawn(move || {
+                let runtime = match runtime::Builder::new_current_thread().enable_all().build() {
+                    Ok(runtime) => Arc::new(runtime),
+                    Err(error) => {
+                        runtime_sender.send(None).expect(
+                            "successful thread creation should always wait for runtime_sender",
+                        );
+                        return Err(StartError::RuntimeNotStarted(error));
+                    }
+                };
+
+                let inner_runtime = runtime.clone();
+                runtime.block_on(async move {
+                    let (databases_sender, databases_receiver) = channel(1);
+                    runtime_sender
+                        .send(Some((inner_runtime, databases_sender.clone())))
+                        .expect("successful thread creation should always wait for runtime_sender");
+
+                    let acceptor_task = acceptor(keep_running.clone(), databases_sender);
+                    let databases_task = router(databases_receiver);
+
+                    keep_running.cancelled().await;
+
+                    let acceptor_task_result = acceptor_task.await;
+                    let databases_task_result = databases_task.await;
+
+                    let mut task_errors = vec![];
+                    if let Err(error) = acceptor_task_result.expect("failed to join acceptor task")
+                    {
+                        task_errors.push(StartError::AcceptError(error));
+                    }
+                    databases_task_result.expect("failed to join databases task");
+
+                    if task_errors.is_empty() {
+                        Ok(())
+                    } else {
+                        Err(StartError::TasksErrored(task_errors))
```

Comment:
Yes, correct. I don't know exactly why this code exists. Maybe for preparation to add a `stop()` method to the `Server` for graceful shutdown.

Reply:
I would suggest joining the join handle immediately before the `Ok(Self { .. })` return to at least log them on server shutdown.

Reply:
I don't know what you mean exactly. We cannot join the thread because then we would wait for its termination. This only works in the error case but we cannot detect this from the main thread. My idea was to store the join handle and join it in a new `stop()` method. Then we would get the error once we are terminating the server.

Reply:
Ah, I see.
Maybe just log the errors here instead of returning?
Or just log them when they occur? Idk why we're even passing these around...
What's the point of knowing that some task failed an hour ago when the server is stopped?

Reply:
There is no benefit in printing them because we are blocked by the cancellation token. Therefore the errors could be printed at the end but then we could just implement correct thread joining. I implemeneted the latter.